lib/backend/postgres/tx.go

```diff
@@ -0,0 +1,470 @@
+		tx.rollback(trace.BadParameter("[BUG] Rollback called with nil error."))
+		tx.cfg.Log.Debug(tx.err)
+
+	case err == nil && tx.err != nil:
+		// Tx in failed state. Rollback called with nil error. Override error to signal issue.
+		tx.err = trace.BadParameter("[BUG] Rollback called with nil error on failed transaction.")
+		tx.cfg.Log.Debug(tx.err)
+	}
+	return trace.Wrap(tx.err)
+}
+
+// rollback and return true when err is not nil.
+//
+// This method is a control-flow helper for the other pgTx methods. Accepting
+// nil errors and returning true or false improves readability in the same
+// manner as trace.Wrap. Its name is purposefully short and discreet.
+func (tx *pgTx) rollback(err error) bool {
+	if err == nil {
+		return false
+	}
+	if e := tx.sqlTx.Rollback(); e != nil {
+		tx.cfg.Log.WithError(err).Errorf("Failed to rollback: %v", e)
+	}
+	tx.err = convertError(err)
+	return true
+}
+
+// DeleteEvents inclusively before an event ID.
+func (tx *pgTx) DeleteEvents(beforeEventID int64) {
+	if tx.err != nil {
+		return
+	}
+
+	const query = `DELETE FROM event WHERE eventid <= $1`
+	_, err := tx.sqlTx.ExecContext(tx.ctx, query, beforeEventID)
+	tx.rollback(err)
+}
+
+// DeleteExpiredLeases removes leases whose expires column is not null and is
+// less than the current time.
+func (tx *pgTx) DeleteExpiredLeases() {
+	if tx.err != nil {
+		return
+	}
+
+	const query = `DELETE FROM lease WHERE (expires IS NOT NULL AND expires < $1)`
+	_, err := tx.sqlTx.ExecContext(tx.ctx, query, tx.now())
+	tx.rollback(err)
+}
+
+// DeleteItems inclusively before an event ID. Items still referencing valid
+// leases or events that have yet to be emitted are excluded.
+func (tx *pgTx) DeleteItems(beforeEventID int64) {
+	if tx.err != nil {
+		return
+	}
+
+	const query = `
+		DELETE FROM item WHERE (key, id) NOT IN (
+			SELECT key, id FROM lease WHERE (expires IS NULL OR expires > $1)
+			UNION
+			SELECT key, id FROM event WHERE eventid >= $2
+		)`
+	_, err := tx.sqlTx.ExecContext(tx.ctx, query, tx.now(), beforeEventID)
+	tx.rollback(err)
+}
+
+// DeleteLease by key returning the backend item ID from the deleted lease.
+// Zero is returned when the delete fails.
+func (tx *pgTx) DeleteLease(key []byte) int64 {
+	if tx.err != nil {
+		return 0
+	}
+
+	var id int64
+	const query = `DELETE FROM lease WHERE key=$1 RETURNING id`
+	err := tx.sqlTx.QueryRowContext(tx.ctx, query, key).Scan(&id)
+	if tx.rollback(err) {
+		return 0
+	}
+	return id
+}
+
+// DeleteLeaseRange removes all leases inclusively between startKey
+// and endKey. It returns the set of backend items deleted. The returned
+// items include only Key and ID.
+func (tx *pgTx) DeleteLeaseRange(startKey, endKey []byte) []backend.Item {
+	if tx.err != nil {
+		return nil
+	}
+
+	const query = `DELETE FROM lease WHERE key >= $1 AND key <= $2 RETURNING key, id`
+	rows, err := tx.sqlTx.QueryContext(tx.ctx, query, startKey, endKey)
+	if tx.rollback(err) {
+		return nil
+	}
+	defer rows.Close()
```

Comment:
Here and elsewhere: error from `rows.Close()` is ignored. Is it fine?

Reply:
Removed. It isn't required. pgx lib calls it.
https://github.com/jackc/pgx/issues/274#issuecomment-300322271
https://github.com/jackc/pgx/blob/v4/rows.go